src/common/admin_socket.cc

```diff
@@ -693,6 +695,234 @@ class GetdescsHook : public AdminSocketHook {
+      return fmt::format(
+        "pending signal ({}) due in {}", 
+        strsignal(signal), 
+        duration_cast<seconds>(remaining));
+    }
+
+    bool cancel()
+    {
+      int wstatus;
+      if (0 == waitpid(pid, &wstatus, WNOHANG)) {
+        kill(pid, SIGKILL);
+        while (pid == waitpid(pid, &wstatus, 0)) {
+          if (WIFEXITED(wstatus)) {
+            return false;
+          }
+          if (WIFSIGNALED(wstatus)) {
+            return true;
+          }
+        }
+      }
+      return false;
+    }
+
+    static std::optional<Killer> fork(int signal_to_send, double delay) {
+      pid_t recipient = getpid();
+      auto until = ceph::coarse_mono_clock::now() + ceph::make_timespan(delay);
+
+      int fresult = ::fork();
+      if (fresult < 0) {
+        return std::nullopt;
+      }
+
+      if (fresult) {
+        // this is parent
+        return {{fresult, signal_to_send, until}};
+      }
+
+      const auto poll_interval = ceph::make_timespan(0.1);
+      auto remaining = (until - ceph::coarse_mono_clock::now());
+      do {
+        using std::chrono::duration_cast;
+        using std::chrono::nanoseconds;
+        std::this_thread::sleep_for(duration_cast<nanoseconds>(std::min(remaining, poll_interval)));
+        if (getppid() != recipient) {
+          // suicide if my parent has changed
+          // this means that the original parent process has terminated
+          _exit(1);
+        }
+        remaining = (until - ceph::coarse_mono_clock::now());
+      } while (remaining > ceph::signedspan::zero());
+
+      kill(recipient, signal_to_send);
+      _exit(0);
+      return std::nullopt;
+    }
+  };
+
+  std::optional<Killer> killer;
+
+  int parse_signal(std::string&& sigdesc, Formatter* f, std::ostream& errss)
+  {
+    int result = 0;
+    std::transform(sigdesc.begin(), sigdesc.end(), sigdesc.begin(),
+        [](unsigned char c) { return std::toupper(c); });
+    if (sigdesc.starts_with("-")) {
+      sigdesc.erase(0, 1);
+    }
+    if (sigdesc.starts_with("SIG")) {
+      sigdesc.erase(0, 3);
+    }
+
+    if (sigdesc == "L") {
+      f->open_object_section("known_signals");
+      for (auto& [name, num] : known_signals) {
+        f->dump_int(name, num);
+      }
+      f->close_section();
+    } else {
+      try {
+        result = std::stoi(sigdesc);
+        if (result < 1 || result > 64) {
+          errss << "signal number should be an integer in the range [1..64]" << std::endl;
+          return -EINVAL;
+        }
+      } catch (std::invalid_argument) {
+        auto sig_it = known_signals.find(sigdesc);
+        if (sig_it == known_signals.end()) {
+          errss << "unknown signal name; use -l to see recognized names" << std::endl;
+          return -EINVAL;
+        }
+        result = sig_it->second;
+      }
+    }
+    return result;
+  }
+
+public:
+  static const char* get_cmddesc()
+  {
+    return "raise "
+           "name=signal,type=CephString,req=false "
```

Comment:
Should not at least `--signal` be required?

Reply:
no, because of the new `--cancel` option. in the body of the `call` method, there are checks and a corresponding error message if the command is invalid.

Reply:
ok